### PR TITLE
add templates alignment node

### PR DESCRIPTION
**Pull request by ShevchenkoRostyslav**
# Add Hamming Distance-Based Template Alignment for Multiframe Aggregation

## PR description

Implements a new template alignment node that aligns iris templates using Hamming distance before aggregation, improving the quality of multiframe template fusion.

## Key Features
- **Hamming Distance Alignment**: Finds optimal rotation offsets to minimize Hamming distance between templates
- **Flexible Reference Selection**: 
  - Use first template as reference (`use_first_as_reference=True`)
  - Use template with minimum sum of distances (`use_first_as_reference=False`)
- **Configurable Parameters**:
  - `rotation_shift`: Maximum rotation range (default: 15)
  - `normalise`: Use normalized Hamming distance (default: False)
- **Multi-wavelet Support**: Handles templates with multiple filter responses

### New Components
- `src/iris/nodes/templates_alignment/hamming_distance_based.py` - Main alignment algorithm
- Added to `iris.HammingDistanceBasedAlignment` in package exports
- Integrated as first step in `multiframe_aggregation_pipeline.yaml`

### Pipeline Integration
```yaml
- name: templates_alignment
  algorithm:
    class_name: iris.HammingDistanceBasedAlignment
    params:
      rotation_shift: 15
      use_first_as_reference: true
      normalise: false
```

### Algorithm Flow
1. Select reference template (first or optimal)
2. For each template, find optimal rotation offset using Hamming distance
3. Apply rotation to align templates
4. Pass aligned templates to aggregation step

## Testing
- **Unit Tests**: 14 test cases covering all functionality, edge cases, and parameter validation
- **E2E Tests**: 13 test cases with realistic data and integration scenarios
- **Coverage**: All methods and parameters tested with mocking and real data

## Limitations
None

## Type

- [x] Feature
- [ ] Refactoring
- [ ] Bugfix
- [ ] DevOps
- [ ] Testing

## Checklist

- [x] I've made sure that my code works as expected by writing unit tests.
- [x] I've checked if my code doesn't generate warnings or errors.
- [x] I've performed a self-review of my code.
- [x] I've made sure that my code follows the style guidelines of the project.
- [x] I've commented hard-to-understand parts of my code.
- [x] I've made appropriate changes in the documentation.

### CODE REVIEW

File: src/iris/nodes/templates_alignment/hamming_distance_based.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        """HammingDistanceBasedAlignment parameters."""
+
+        rotation_shift: conint(ge=0, strict=True) = Field(
+            default=15, description="Maximum rotation shift allowed for alignment"
+        )
+        use_first_as_reference: bool = Field(
+            default=True,
+            description="Use first template as reference, otherwise use template with minimal sum of distances",
+        )
+        normalise: bool = Field(default=False, description="Whether to use normalized Hamming distance for alignment")
```
Comment (TanguyJeanneau) [inline src/iris/nodes/templates_alignment/hamming_distance_based.py:47]: Let's normalize the hamming distance by default too, with 15 rotations we'll start to have quite some masked nits 🙂

File: src/iris/nodes/templates_alignment/hamming_distance_based.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            rotated_mask_code = np.roll(mask_code, rotation, axis=1)
+
+            rotated_iris_codes.append(rotated_iris_code)
+            rotated_mask_codes.append(rotated_mask_code)
+
+        return IrisTemplate(
+            iris_codes=rotated_iris_codes,
+            mask_codes=rotated_mask_codes,
+            iris_code_version=template.iris_code_version,
+        )
```
Comment (ycbiometrics) [inline src/iris/nodes/templates_alignment/hamming_distance_based.py:196]: Shouldn't we compare the HD with a threshold to make sure only likely genuine samples are being aggregated? In addition, do you only consider frames that pass our quality validators or are you taking any frames? If the latter, then we should also make sure the match has enough overlap. This can help prevent to blend multiple identities (or eyes) or heavily occluded (noisy) frames into the aggregated template.

Comment (ycbiometrics) [reply src/iris/nodes/templates_alignment/hamming_distance_based.py:196]: It will be great to add some test cases for these scenarios.

File: src/iris/nodes/templates_alignment/hamming_distance_based.py (hunk 1 above)
Comment (ycbiometrics) [reply src/iris/nodes/templates_alignment/hamming_distance_based.py:47]: yes, I agree we should used normalized HD.

**Review by ycbiometrics — COMMENTED** [review]
I have some comments.

File: src/iris/nodes/templates_alignment/hamming_distance_based.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    class Parameters(Algorithm.Parameters):
+        """HammingDistanceBasedAlignment parameters."""
+
+        rotation_shift: conint(ge=0, strict=True) = Field(
+            default=15, description="Maximum rotation shift allowed for alignment"
+        )
+        use_first_as_reference: bool = Field(
+            default=True,
+            description="Use first template as reference, otherwise use template with minimal sum of distances",
```
[the remark this answers is not in the document]
Comment (ShevchenkoRostyslav) [reply src/iris/nodes/templates_alignment/hamming_distance_based.py:45]: @TanguyJeanneau i made it False by default to reduce the overall runtime as it's currently implemented in n**2 time complexity. 
No particular reason behind the min sum tbh. Why would we square it though? 
- linear is simple and easy to understand and debug
- HD is already normalized and bounded
- linear is not so sensitive to some weird outliers (e.g. if we in the future will combine tempaltes from different signups that might have some "far apart" frames) 
But again - we could make it as we want. I can also make it configurable: "linear" / "root_mean_squared" ...

File: src/iris/nodes/templates_alignment/hamming_distance_based.py (hunk 2 above)
Comment (ycbiometrics) [reply src/iris/nodes/templates_alignment/hamming_distance_based.py:238]: I must have missed the use case. Why can the frames be from different people if someone wants it? I am just afraid we cannot rely on the user to ensure only good quality genuine samples are being aggregated. For example, we have seen orbs move around and pick up right eye as left eye or pick out someone else's eye when multiple people are in the view. The risk is that once a different eye is blended in, it will be very hard to detect or reverse.
